packages/next/src/server/route-modules/app-route/module.ts

```diff
@@ -388,11 +399,187 @@ export class AppRouteRouteModule extends RouteModule<
+                       * and dynamicIO is on we follow a similar pattern to rendering.
+                       *
+                       * We first run the handler letting caches fill. If something synchronously dynamic occurs
+                       * during this prospective render then we can infer it will happen on every render and we
+                       * just bail out of prerendering.
+                       *
+                       * Next we run the handler againg and we check if we get a result back in a microtask.
+                       * Next expects the return value to be a Response or a Thenable that resolves to a Response.
+                       * Unfortunately Response's do not allow for acessing the response body synchronously or in
+                       * a microtask so we need to allow one more task to unwrap the response body. This is a slightly
+                       * different semantic than what we have when we render and it means that certain tasks can still
+                       * execute before a prerender completes such as a carefully timed setImmediate.
+                       *
+                       * Functionally though IO should still take longer than the time it takes to unwrap the response body
+                       * so our heuristic of excluding any IO should be preserved.
+                       */
+                      let prospectiveRenderIsDynamic = false
+                      const cacheSignal = new CacheSignal()
+                      let dynamicTracking =
+                        createDynamicTrackingState(undefined)
+                      const prospectiveRoutePrerenderStore: PrerenderStore = {
+                        cacheSignal,
+                        // During prospective render we don't use a controller
+                        // because we need to let all caches fill.
+                        controller: null,
+                        dynamicTracking,
+                      }
+                      let prospectiveResult
+                      try {
+                        prospectiveResult = this.prerenderAsyncStorage.run(
+                          prospectiveRoutePrerenderStore,
+                          handler,
+                          request,
+                          {
+                            params: context.params
+                              ? parsedUrlQueryToParams(context.params)
+                              : undefined,
+                          }
+                        )
+                      } catch (err) {
+                        if (isPrerenderInterruptedError(err)) {
+                          // the route handler called an API which is always dynamic
+                          // there is no need to try again
+                          prospectiveRenderIsDynamic = true
+                        }
+                      }
+                      if (
```

Comment:
we already have a `isThenable` helper could be used

Reply:
That's buried in some reducer code. We could extract it but I don't want to pull it into this server app route module